gadgets/src/poseidon/mod.rs

```diff
@@ -0,0 +1,438 @@
+#![allow(non_snake_case)]
+
+use crate::poseidon::builder::Poseidon;
+use curve25519_dalek::scalar::Scalar;
+use bulletproofs::r1cs::{ConstraintSystem, R1CSError, Prover, Verifier};
+use bulletproofs::{PedersenGens};
+use bulletproofs::r1cs::LinearCombination;
+use crate::utils::{AllocatedScalar, constrain_lc_with_scalar};
+
+
+
+pub mod sbox;
+pub use sbox::*;
+pub mod builder;
+pub use builder::*;
+
+#[cfg(test)]
+pub mod tests;
+
+
+fn Poseidon_permutation(
```

Comment:
This func is correct, but it could be cleaned up a little bit. e.g. all rounds loops can be put into one, mix layer can be put into a separate function

Same goes for `Poseidon_permutation_constraints`

Reply:
fair, I'm not so sure we'll need to make it more concise since this is fairly readable against the Sage implementation.